File: java/leetcode_grind/Day569.java
package leetcode_grind;

import java.util.Arrays;
import java.util.HashMap;

public class Day569 {
    // https://leetcode.com/problems/hand-of-straights/description
    static class Solution1 {
        public boolean isNStraightHand(int[] hand, int groupSize) {
            var cnt = new HashMap<Integer, Integer>();
            for (var v : hand) {
                cnt.put(v, cnt.getOrDefault(v, 0) + 1);
            }
            Arrays.sort(hand);
            for (var ptr = 0; ptr < hand.length; ptr++) {
                if (cnt.getOrDefault(hand[ptr], 0) == 0) {
                    continue;
                }
                for (var j = hand[ptr]; j < hand[ptr] + groupSize; j++) {
                    if (cnt.getOrDefault(j, 0) > 0) {
                        cnt.put(j, cnt.get(j) - 1);
                        // System.out.printf("Take j %s \n", j);
                    } else {
                        // System.out.printf("No candidate j %s found, exit \n", j);
                        return false;
                    }
                }
                // System.out.printf("End forming group \n");
            }
            return true;
        }
    }

    // https://leetcode.com/problems/hand-of-straights/description
    static class Solution2 {
        public boolean isNStraightHand(int[] hand, int groupSize) {
            if (hand.length % groupSize != 0) {
                return false;
            }

            HashMap<Integer, Integer> cardCount = new HashMap<>();
            for (int card : hand) {
                int count = cardCount.getOrDefault(card, 0);
                cardCount.put(card, count + 1);
            }

            for (int card : hand) {
                int startCard = card;
                while (cardCount.getOrDefault(startCard - 1, 0) > 0) {
                    startCard--;
                }
                while (startCard <= card) {
                    while (cardCount.getOrDefault(startCard, 0) > 0) {
                        for (int nextCard = startCard; nextCard < startCard + groupSize; nextCard++) {
                            if (cardCount.getOrDefault(nextCard, 0) == 0) {
                                return false;
                            }
                            cardCount.put(nextCard, cardCount.get(nextCard) - 1);
                        }
                    }
                    startCard++;
                }
            }
            return true;
        }
    }
}

File: python/leetcode_grind/day_569.py
# https://leetcode.com/problems/hand-of-straights/description
class Solution1:
    def isNStraightHand(self, hand: List[int], groupSize: int) -> bool:
        hand_size = len(hand)
        if hand_size % groupSize != 0:
            return False

        card_count = Counter(hand)
        min_heap = list(card_count.keys())
        heapq.heapify(min_heap)

        while min_heap:
            current_card = min_heap[0]
            for i in range(groupSize):
                if card_count[current_card + i] == 0:
                    return False
                card_count[current_card + i] -= 1
                if card_count[current_card + i] == 0:
                    if current_card + i != heapq.heappop(min_heap):
                        return False
        return True

# https://leetcode.com/problems/search-in-rotated-sorted-array/description/
class Solution2:
    def search(self, nums: List[int], target: int) -> int:
        n = len(nums)
        left, right = 0, n-1
        while left <= right:
            mid = left + (right - left) // 2
            if nums[mid] == target:
                return mid
            elif nums[mid] >= nums[left]:
                if target >= nums[left] and target < nums[mid]:
                    right = mid - 1
                else:
                    left = mid + 1
            else:
                if target <= nums[right] and target > nums[mid]:
                    left = mid + 1
                else:
                    right = mid - 1
        return -1